sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -442,4 +442,14 @@ object QueryPlan {
         }
     }.canonicalized.asInstanceOf[T]
   }
+
+  /** Normalize and reorder the expressions in the given sequence. */
```

Comment:
It links predicates with `AND` and pass it to `def normalizeExprId`, in which `Canonicalize.execute` is called. That's where we do the reordering.